clients/callback_client.go

```diff
@@ -0,0 +1,111 @@
+package clients
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const max_retries = 3
+const backoff_millis = 200
+
+type CallbackClient struct {
+	httpClient *http.Client
+}
+
+func NewCallbackClient() CallbackClient {
+	return CallbackClient{
+		httpClient: &http.Client{},
+	}
+}
+
+func (c CallbackClient) DoWithRetries(r *http.Request) error {
+	var resp *http.Response
+	var err error
+	for x := 0; x < max_retries; x++ {
+		resp, err = c.httpClient.Do(r)
+		if err == nil && resp.StatusCode == http.StatusOK {
+			return nil
+		}
+
+		// Back off to give us more chance of succeeding during a network blip
+		time.Sleep(backoff_millis * time.Millisecond)
```

Comment:
So, in total we wait: 200 + 400 + 600 = 1.2s. Is it shorter than the frequency of the updates? I think if it's not possible that callback receives the status not in order.

Reply:
That's a great point, I think we should handle out-of-order updates gracefully anyway (since they could happen even without this backoff). I'll add a timestamp field to the messages we send.

Reply:
cc @victorges - something to bear in mind for you too

Reply:
Also @leszko, it should just be ~600ms total - I kept it simple and didn't try to do an exponential backoff

Reply:
Added timestamps now. I think Studio will probably still need to have some level of "state machine" knowledge though, at least in terms of terminal states (i.e if it gets an in-progress notifiction after an error/completion one, it can ignore it)

Reply:
@leszko also consider special cases: 
- Target machine is down 
- DNS lookup takes some time if not in cache
- If target server is overloaded and kernel accepts connection there is no automatic timeout to produce response

Actual connect may take minutes before error is returned. On my box sometimes 3 seconds sometimes 130 seconds:
```
$ ./test_connect 10.15.208.3 443
Connect error DNS=0 connect=3.04924
$ ./test_connect 192.168.1.2 443
Connect error DNS=0 connect=130.417
```

Maybe we should always use timeout on our side.

Reply:
Great point @AlexKordic, I've added a timeout of 5s to the HTTP Client now